kong/plugins/prometheus/wasmx.lua

```diff
@@ -0,0 +1,210 @@
+local buffer = require "string.buffer"
+local wasm = require "kong.runloop.wasm"
+local wasmx_shm
+
+
+local fmt = string.format
+local str_find = string.find
+local str_sub = string.sub
+local table_insert = table.insert
+local table_sort = table.sort
+local buf_new = buffer.new
+local ngx_say = ngx.say
+
+
+local _M = {}
+
+
+local FLUSH_EVERY = 100
+
+
+local function sorted_iter(ctx, i)
+  i = i + 1
+
+  local v = ctx.t[ctx.sorted_keys[i]]
+
+  if v ~= nil then
+    return i, v
+  end
+end
+
+
+local function sorted_pairs(t)
+  local sorted_keys = {}
+
+  for k, _ in pairs(t) do
+    table_insert(sorted_keys, k)
+  end
+
+  table_sort(sorted_keys)
+
+  return sorted_iter, { t = t, sorted_keys = sorted_keys }, 0
+end
+
+--
+-- Convert a pw_key into a pair of metric name and labels
+--
+-- pw_key follows the form `pw.<filter_name>.<metric_name>`
+-- `<metric_name>` might contain labels, e.g. a_metric_label1="v1";
+-- if it does, the position of the first label corresponds to the end of the
+-- metric name and is used to discard labels from <metric_name>.
+local function parse_pw_key(pw_key)
+  local metric_name = pw_key
+  local _, _, filter_name = str_find(pw_key, "pw.(%a+)%.")
+  local filter_meta = wasm.filter_meta[filter_name] or {}
+  local label_patterns = {}
+  local labels = {}
+  local first_label_pos = #pw_key
+
+  if filter_meta.metrics and filter_meta.metrics.label_patterns then
+    label_patterns = filter_meta.metrics.label_patterns
+  end
+
+  for _, pair in ipairs(label_patterns) do
+    local label_pos, _, _, value = str_find(pw_key, pair.pattern)
+
+    if label_pos and value then
+      table_insert(labels, { pair.label, value })
+
+      first_label_pos = (label_pos < first_label_pos) and label_pos or first_label_pos
+    end
+  end
+
+  if first_label_pos ~= #pw_key then
+    -- discarding labels from metric_name
+    metric_name = str_sub(pw_key, 1, first_label_pos - 1)
+  end
+
+  return metric_name, labels
+end
+
+
+local function parse_key(key)
+  -- TODO: parse wa. (WasmX metrics) and lua. (metrics defined in Lua land)
+  local header = { pw = "pw." }
+
+  local name = key
+  local labels = {}
+
+  local is_pw = #key > #header.pw and key:sub(0, #header.pw) == header.pw
+
+  if is_pw then
+    name, labels = parse_pw_key(key)
+  end
+
+  name = name:gsub("%.", "_")
+
+  return name, labels
+end
+
+
+local function serialize_labels(labels)
+  local buf = buf_new()
```

Comment:
There are a few places in this file (here and in `serialize_metric()`) where short-lived string buffers are being created within a loop. I would suggest taking a pass at the code to see where said buffers can be constructed in an outer scope and shared/reused. Metrics collection is a hot code path, so frugality is paramount.

```suggestion
local function serialize_labels(labels, buf)
```